Codejam Virtual Keyboard

alexeikravchuk opened the pull request:
1. задание - https://github.com/rolling-scopes-school/tasks/blob/master/tasks/codejam-virtual-keyboard.md
2. скриншот - https://imgur.com/IlkKE6n
3. демо - https://alexei-keyboard.netlify.com
4. дата сдачи: 2020-04-03 / дата дедлайна: 2020-04-07 23:59

5. Total 110 баллов
Task ( 8 / 8 )

`минимальный набор:` (2 / 2) 30 баллов
+ реализована генерация DOM-элементов и `body` в index.html пустой изначально (может находится только тег `script`): `+20`
+ нажатие на кнопку на физической клавиатуре подсвечивает кнопку на виртуальной (проверять следует нажатие цифр, букв, знаков препинания, `backspace`, `del` (если она присутствует), `enter`, `shift`, `alt`, `ctrl`, `tab`, `caps lock`, `space`, стрелки вниз-вверх-влево-вправо): `+10`

`стандартный набор:` (2 / 2) 30 баллов
+ есть переключение между русским и английским языком (сочетание клавиш для переключения языка должно быть указано на странице с клавиатурой), а также сохранение выбранного языка: `+15`
+ клики мышкой по кнопкам на виртуальной клавиатуре или нажатие на кнопки физической клавиатуры, выводят символы в инпут (textarea): `+15`

`дополнительный набор:` (1 / 1) 15 баллов
+ реализована анимация нажатия на кнопку: `+15`

`технические требования:` (3 / 3) 35 баллов
+ использование в коде фишек стандарта ES6 и выше (classes, деструктуризация и т.д.): `+15`
+ использование eslint: `+10`
+ требования к репозиторию, коммитам и PR выполнены: `+10`

`штрафы:` 0 баллов
- ошибки в консоли, связанные с исполняемым кодом (ошибки типа `favicon.ico: Failed to load resource: the server responded with a status of 404` не учитываются) или предупреждения eslint-config-airbnb-base: `-15`

alexeikravchuk committed:
feat: load demo version

alexeikravchuk committed:
feat: prod demo deploy

alexeikravchuk committed:
Merge pull request #1 from alexeikravchuk/codejam

RSS Virtual Keyboard

alexeikravchuk committed:
Update README.md

alexeikravchuk committed:
feat: added arrow action

alexeikravchuk committed:
fix: eslint-config-airbnb-base requirements met

alexeikravchuk committed:
feat: load develop project

alexeikravchuk committed:
refactor: load prod files

alexeikravchuk committed:
docs: update readme with additional information

alexeikravchuk committed:
refactor: eslint fix

alexeikravchuk committed:
Merge branch 'gh-pages' of https://github.com/alexeikravchuk/virtual-keyboard into gh-pages

alexeikravchuk committed:
fix: fixed some bugs

alexeikravchuk committed:
fix: fixed bug (backspace in start line)

alexeikravchuk committed:
fix: changin language with right shift

klimenkomaksim commented on src/index.js:
```diff
@@ -0,0 +1,40 @@
+  wrapper.append(textarea);
+
+  const keyboardWrapper = document.createElement('div');
+  keyboardWrapper.classList.add('keyboard', 'keyboard_wrapper');
```
could you please move creation and adding classes to separate func to avoid copypasta

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,176 @@
+  keyAction(e) {
+    e.preventDefault();
+    const keys = this.container.querySelectorAll('.key');
+    keys.forEach((key) => {
```
you can convert `NodeList` with `Array.from` and then use methods from `Array.prototype` to simplify

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,176 @@
+      const code = key.getAttribute('data-code');
+      if (code === e.code) {
+        if (e.type === 'keydown') {
+          this.keyDownAction(key);
```
would be better do manipulations with the needed key outside of find func

alexeikravchuk committed:
fix: load index.html with links to js, css

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,176 @@
+    }
+    if (e.type === 'mouseup') {
+      const keys = this.container.querySelectorAll('.key');
+      keys.forEach((key) => {
```
I think you could just pass `this.keyUpAction` as callback

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,176 @@
+
+  keyDownAction(key) {
+    if (key.dataset.code !== 'CapsLock') {
+      key.classList.add('active');
```
what will happen if you use `toggle` in this case too?

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,176 @@
+
+    if (!key.classList.contains('special_key')) {
+      this.model.addCharacterToLine(key.dataset.code);
+    } else {
```
https://www.itamarweiss.com/personal/2018/02/28/return-early-pattern.html

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,176 @@
+        this.isControlLeftOn = true;
+        this.changeLanguageCheck();
+        break;
+      case 'ControlRight':
```
why changing language possible only with left ctrl?

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,176 @@
+        this.model.toggleUpperCase(true);
+        break;
+      case 'ShiftRight':
+        this.changeLanguageCheck();
```
you can move the logic for shift key down to separate func 

that will help to avoid copypasta and to uprise readability

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,176 @@
+        break;
+      case 'ShiftLeft':
+        this.isShiftLeftOn = true;
+        this.changeLanguageCheck();
```
it's a good practice to name methods with verbs

klimenkomaksim commented on src/js/Model.js:
```diff
@@ -0,0 +1,129 @@
+const specialKeysJSON = '{"Backspace":{"id":13,"value":"backspace","width":"double_quarter"},"Tab":{"id":14,"value":"tab","width":"one_quarter"},"Delete":{"id":28,"value":"del","width":"plus_quarter"},"CapsLock":{"id":29,"value":"caps-lock","width":"one_half"},"Enter":{"id":41,"value":"enter","width":"double_quarter"},"ShiftLeft":{"id":42,"value":"shift","width":"double_quarter"},"ShiftRight":{"i ...
+const characterKeysJSON = '{"Backquote":{"id":0,"value":{"en":["`","~"],"ru":["ё","Ё"]}},"Digit1":{"id":1,"value":{"en":["1","!"],"ru":["1","!"]}},"Digit2":{"id":2,"value":{"en":["2","@"],"ru":["2","\\""]}},"Digit3":{"id":3,"value":{"en":["3","#"],"ru":["3","№"]}},"Digit4":{"id":4,"value":{"en":["4","$"],"ru":["4",";"]}},"Digit5":{"id":5,"value":{"en":["5","%"],"ru":["5","%"]}},"Digit6":{"id":6," ...
```
why it's here if you have `character-keys.json` & `special-keys.json` files?

alexeikravchuk committed:
refactor: creating elements in a separate function

alexeikravchuk committed:
refactor: creating elements in a separate function

alexeikravchuk committed:
docs: change the link to the site

alexeikravchuk committed:
Merge branch 'gh-pages' of https://github.com/alexeikravchuk/virtual-keyboard into gh-pages

alexeikravchuk committed:
refactor: keyAction functions

+defined new function determineKeyAction();
+ convertet NodeList with Array.from and then used methods from Array.prototype

alexeikravchuk committed:
refactor: language change algorithm

alexeikravchuk committed:
refactor: simplified method specialKeyUpAction

alexeikravchuk committed:
refactor: loading keyboard keys from json

solved the problem of webpack with async functions (got an error "regeneratorRuntime is not defined")

klimenkomaksim commented on src/js/View.js:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  drawKeyboard(numberOfKeys, specialKeys) {
+    const arrowsUpDown = document.createElement('div');
```
please, use you own func `createElement`

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,132 @@
+  keyAction(e) {
+    e.preventDefault();
+
+    function determineKeyAction(key) {
```
why you create this func inside other func?

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,132 @@
+    if (targetKey) determineKeyAction.call(this, targetKey);
+  }
+
+  mouseAction(e) {
```
fix as discussed

klimenkomaksim commented on src/js/Controller.js:
```diff
@@ -0,0 +1,132 @@
+    if (key.classList.contains('special_key')) {
+      this.specialKeyUpAction(key);
+    }
+    if (this.lastPressedKey.length < 1) this.lastPressedKey.push(null);
```
why is that?

alexeikravchuk committed:
refactor: implement return-early-pattern

alexeikravchuk committed:
refactor: used own func createElement

single-line expression after if placed in block

alexeikravchuk commented:
all code comments corrected
